bigquery/google/cloud/bigquery/_pandas_helpers.py

```diff
@@ -176,5 +207,163 @@ def to_parquet(dataframe, bq_schema, filepath):
+
+    read_options = bigquery_storage_v1beta1.types.TableReadOptions()
+    if selected_fields is not None:
+        for field in selected_fields:
+            read_options.selected_fields.append(field.name)
+
+    requested_streams = 0
+    if preserve_order:
+        requested_streams = 1
+
+    session = bqstorage_client.create_read_session(
+        table.to_bqstorage(),
+        "projects/{}".format(project_id),
+        read_options=read_options,
+        requested_streams=requested_streams,
+    )
+
+    # We need to parse the schema manually so that we can rearrange the
+    # columns.
+    schema = json.loads(session.avro_schema.schema)
+    column_names = [field["name"] for field in schema["fields"]]
+
+    # Avoid reading rows from an empty table. pandas.concat will fail on an
+    # empty list.
+    if not session.streams:
+        yield pandas.DataFrame(columns=column_names)
+        return
+
+    total_streams = len(session.streams)
+
+    # Use _DownloadState to notify worker threads when to quit.
+    # See: https://stackoverflow.com/a/29237343/101923
+    download_state = _DownloadState()
+
+    # Create a queue to collect frames as they are created in each thread.
+    worker_queue = queue.Queue()
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=total_streams) as pool:
+        try:
+            # Manually submit jobs and wait for download to complete rather
+            # than using pool.map because pool.map continues running in the
+            # background even if there is an exception on the main thread.
+            # See: https://github.com/googleapis/google-cloud-python/pull/7698
+            not_done = [
+                pool.submit(
+                    _download_dataframe_bqstorage_stream,
+                    download_state,
+                    bqstorage_client,
+                    column_names,
+                    dtypes,
+                    session,
+                    stream,
+                    worker_queue,
+                )
+                for stream in session.streams
+            ]
+
+            while not_done:
+                done, not_done = _nowait(not_done)
+                for future in done:
+                    # Call result() on any finished threads to raise any
+                    # exceptions encountered.
+                    future.result()
+
+                try:
+                    frame = worker_queue.get(timeout=_PROGRESS_INTERVAL)
+                    yield frame
+                except queue.Empty:  # pragma: NO COVER
+                    continue
+
+            # Return any remaining values after the workers finished.
+            while not worker_queue.empty():
+                try:
+                    # Include a timeout because even though the queue is
+                    # non-empty, it doesn't guarantee that a subsequent call to
+                    # get() will not block.
+                    frame = worker_queue.get(timeout=_PROGRESS_INTERVAL)
```

Comment:
Since the only place that _takes_ the items from the queue is here (the workers only maybe-put them in the queue), what's the case of the `get()` call blocking? Isn't the queue always non-empty at this point, meaning that `get()` will never block (or raise `Empty` if timeout is used)?

Reply:
Yes, this is the only thread that reads from the queue. So, since the queue is non-empty, this call should never block.

I'm being conservative here, since the [Python docs for Queue.empty()](https://docs.python.org/3/library/queue.html#queue.Queue.empty) aren't specific about in what cases we can expect `get` to not block.

Reply:
I checked the [source](https://github.com/python/cpython/blob/3.7/Lib/queue.py#L169-L170), it only blocks when the queue is empty. But I'm also fine with conservative approach in general, since an incorrect assumption can result in a much more severe bug than having a few extra lines of code.